test/Runtime/weak-reference.swift

```diff
@@ -0,0 +1,69 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+// REQUIRES: objc_interop
+
+import StdlibUnittest
+
+// Also import modules which are used by StdlibUnittest internally. This
+// workaround is needed to link all required libraries in case we compile
+// StdlibUnittest with -sil-serialize-all.
+import SwiftPrivate
+#if _runtime(_ObjC)
+import ObjectiveC
+#endif
+
+import Dispatch
+
+class Thing {}
+
+class WeakBox<T: AnyObject> {
+  weak var weak: T?
+  init(_ ref: T) { weak = ref }
+}
+
+let WeakReferenceTests = TestSuite("WeakReferenceTests")
+
+WeakReferenceTests.test("thread safety: indirect capture [SR-192]")
+  .skip(.Always("Useful test, though non-deterministic"))
+  .code {
+  let q = dispatch_queue_create("", DISPATCH_QUEUE_CONCURRENT)
+
+  // Capture a weak reference via its container object
+  // "https://bugs.swift.org/browse/SR-192"
+  let iterations = 200_000
+  for i in 1...iterations {
+    let box = WeakBox(Thing())
+    dispatch_async(q) {
```

Comment:
We have race test infrastructure (stdlib/private/StdlibUnittest/RaceTest.swift) that makes it possible to write such tests in a cross-platform way.  Could you use that instead of dispatch?